element: Make createInterpolateElement TS/type smart

manzoorwanijk opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR makes `createInterpolateElement` type/TS smart by interpolating the string passed as the first argument and then providing type-checking for the `conversionMap` provided as the second argument

```ts
import { createInterpolateElement } from '@wordpress/element';

const str =
	'This is a <span>string</span> with <a>a link</a> and a self-closing <CustomComponentA/> tag';
// Example usage demonstrating the enhanced typing:
createInterpolateElement( str, {
	// TypeScript will now enforce that only
	// 'span', 'a', and 'CustomComponentA' keys are allowed
	// and will provide autocompletion for these keys
	span: createElement( 'span' ),
	a: createElement( 'a', { href: 'https://github.com' } ),
	CustomComponentA: createElement( 'div' ),
	// It will not allow keys not present in the string
	CustomComponentB: createElement( 'div' ), // This would cause a TypeScript error
} );

// This will validate that the values are React elements
createInterpolateElement( str, {
	span: createElement( 'span' ),
	a: createElement( 'a' ),
	CustomComponentA: 'not an element', // This would cause a TypeScript error
} );
```

## Why?

To enhance type-safety and allow auto-completions for `createInterpolateElement`

## How?

The PR uses some advanced TypeScrip string interpolation to create types like `ConversionMap` which uses another type `ExtractTags` which extracts tags from the passed string. The types are commented enough to make them readable and maintainable

## Testing Instructions

- Copy the above snippet in your editor
- Confirm that you see the auto-completions
- Confirm that conversion map is type-checked as demonstrated

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<img width="979" height="926" alt="Screenshot 2025-09-05 at 12 19 31 PM" src="https://github.com/user-attachments/assets/cf828f72-939a-4aa7-955a-3c0de9fe4498" />

manzoorwanijk committed:
element: Make createInterpolateElement TS smart

manzoorwanijk committed:
Update docs

manzoorwanijk committed:
Add support for recent i18n type improvements

manzoorwanijk committed:
Improve types for README docs

manzoorwanijk commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,55 @@
+ * - Only allows properties for tags found in the template literal
+ */
+export type ConversionMap< T extends string > = Partial<
+	Record< ExtractTags< T >, React.ReactElement >
```
Fixed

SirLouen commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,57 @@
+
+/**
+ * Helper type to extract tag name and handle closing/self-closing indicators
+ * Matches the tokenizer regex: /<(\/)?(\w+)\s*(\/)?>/g
```
Do you think it is good to add so much specific info to a doc? I would keep it simpler, as despite its unlikelihood, there is an open possibility for change, and it's unlikely that this will be updated afterward).

manzoorwanijk replied:
I think you are right. Let us remove it.

```suggestion
```

manzoorwanijk committed:
Remove unnecessary JSDoc comment

jsnajdr commented on packages/element/src/test/create-interpolate-element.tsx:
The `item` tag is not part of the inferred conversion map type, is that expected? Only `somethingElse` is detected, and the map type is `Partial<Record<"somethingElse">>`.

jsnajdr replied:
One additional unit test that would be nice to have is where we test explicitly if `ExtractTags` really extracts all the tags. For example, from `<a><b><c>`, the type should be `"a"|"b"|"c"`. That would test the template matching algorithm.

manzoorwanijk replied:
Good catch!

jsnajdr commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,56 @@
+ */
+type TranslatableText< T extends string > = string & {
+	readonly __translatableText: T;
+};
```
Can we have an unit test for a `TranslatableText` argument?

My understanding is that we need this special case in order to extract the "real" string from the right place. Not from the actual string itself, but from its `__translatableText` field.

manzoorwanijk replied:
Added a type-level test that verifies `InterpolationString` correctly unwraps a `TranslatableText` input, and `ExtractTags` then extracts the tags from the underlying literal type.

jsnajdr commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,56 @@
+	T extends `${ string }<${ infer Tag }>${ infer After }`
+		? ExtractTagName< Tag > extends never
+			? ExtractTags< After >
+			: ExtractTagName< Tag > | ExtractTags< After >
```
One of the `extends never` check is redundant. We can do:
```ts
T extends `${string}<${infer T}>${infer A}`
  ? ExtractTagName<T> | ExtractTags<A>
  : never
```
That's because `never|T` is always equivalent to `T`, `never` is the TypeScript "bottom type".

manzoorwanijk replied:
Good call — simplified.

jsnajdr reviewed (commented):
Hi @manzoorwanijk :wave: What if we finished this PR? 🙂 I posted some review comments.

manzoorwanijk committed:
Simplify ExtractTags by removing redundant extends never check

The `extends never` branch is unnecessary because `never | T` is
always equivalent to `T` — `never` is TypeScript's bottom type.

manzoorwanijk committed:
Fix ExtractTagName to handle whitespace before self-closing slash

The runtime tokenizer regex `/<(\/)?(\w+)\s*(\/)?>/g` allows optional
whitespace between the tag name and the closing slash (e.g. `<item />`).
The type-level ExtractTagName was incorrectly filtering these as
"tags with spaces". Reorder the checks so `<name />` is recognized
as a valid self-closing tag before the general space filter applies.

manzoorwanijk committed:
Add type-level tests for ExtractTags and TranslatableText

Verify that ExtractTags correctly extracts all tag names from a
template literal string, and that InterpolationString properly
unwraps TranslatableText to extract the underlying literal type.

jsnajdr commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,58 @@
+ */
+type ExtractTagName< T extends string > = T extends `/${ string }`
+	? never // Skip closing tags like "/div"
+	: T extends `${ infer Name } /`
```
This will fail again when there are two spaces. ChatGPT proposed a recursive helper:
```ts
type TrimRightSpaces<S extends string> =
	S extends `${infer Rest} `
		? TrimRightSpaces<Rest>
		: S;

type ExtractTagName<S extends string> =
	// forbid leading space
	S extends ` ${string}`
		? never
		: TrimRightSpaces<S> extends infer Name extends string
			// optional: reject empty tag
			? Name extends ""
				? never
				// forbid inner spaces after trailing spaces were removed
				: Name extends `${string} ${string}`
					? never
					: Name
			: never;
```
This could be used. There is also a nice trick with double `extends` that helps prevent two calls of `TrimRightSpaces`:
```ts
TrimRightSpaces<S> extends infer Name extends string
```

manzoorwanijk replied:
Good catch! Updated.

manzoorwanijk committed:
Use TrimTrailingSpaces for robust whitespace handling in ExtractTagName

Replace the single-space pattern match with a recursive
TrimTrailingSpaces helper that handles any number of trailing
spaces before the self-closing slash, matching the runtime
tokenizer's \s* behavior.

jsnajdr commented on packages/element/src/types.ts:
```diff
@@ -0,0 +1,68 @@
+export type ExtractTags< T extends string > =
+	T extends `${ string }<${ infer Tag }>${ infer After }`
+		? ExtractTagName< Tag > | ExtractTags< After >
+		: never;
```
There's one more little bug that I don't know how to solve. Consider this call where the input is a generic `string` and the conversion map has invalid values (number instead of element):
```ts
createInterpolateElement( foo, { a: 1 } )
```
There will be no type error, although the value of `a` should be reported as wrong. The reason is that for `string` input, conversion map type is `Record<never, ReactElement>`. Because of the `never` key, this is effectively `{}`, and the `a` field is considered an "extra" field that is not typechecked.

A better type would be `Record<string, ReactElement>`. This can be achieved by selectively returning `string` or `never` from `ExtractTags`:
```ts
type ExtractTags<T extends string, R = string> =
  T extends `tag-pattern`
    ? ExtractTagName<Tag> | ExtractTags<After, never>
    : R;
```
This returns `string` when `T` is a generic string, but nested calls will return `never`.

But this is not satisfactory because then the following call:
```ts
createInterpolateElement( 'foo', { a: <em /> } )
```
will stop detecting that `a` is not a valid key. The `foo` literal also doesn't match the `tag-pattern`, just like the generic string. We need to be able to distinguish between a generic `string` and a literal value.

jsnajdr reviewed (approved):
I think this is ready to go :+1:

One more thing that would be nice to solve in a follow-up. Currently the `createInterpolateElement` types are much less useful than they could be, because so many calls use a nested `sprintf`:
```ts
createInterpolateElement( sprintf( '<Name>%1$s</Name>', name ), { Name } );
```
The return type of `sprintf` is `string`, and the smart types infer nothing.

We could use a similar trick as in `TranslatableText`. The `__( 'string to translate' )` helper has the same problem, the result is a generic `string`, but we can pass through the original literal as the `__translatableText` virtual field. The `sprintf` return value could be a similar type.
